givenergy_modbus_async/model/ems.py

```diff
@@ -0,0 +1,207 @@
+from enum import IntEnum, StrEnum
+import math
+
+from pydantic import BaseConfig, create_model
+
+from givenergy_modbus_async.model.register import HR, IR
+from givenergy_modbus_async.model.register import (
+    Converter as C,
+)
+from givenergy_modbus_async.model.register import (
+    RegisterDefinition as Def,
+)
+from givenergy_modbus_async.model.register import (
+    RegisterGetter,
+)
+
+class Generation(StrEnum):
+    """Known Generations"""
+
+    GEN1 = "Gen 1"
+    GEN2 = "Gen 2"
+    GEN3 = "Gen 3"
+
+    @classmethod
+    def _missing_(cls, value: int):
+        """Pick generation from the arm_firmware_version."""
+        arm_firmware_version_to_gen = {
+            3: cls.GEN3,
+            8: cls.GEN2,
+            9: cls.GEN2,
+        }
+        key = math.floor(int(value) / 100)
+        if gen := arm_firmware_version_to_gen.get(key):
+            return gen
+        else:
+            return cls.GEN1
+
+class Model(StrEnum):
+    """Known models of inverters."""
+
+    HYBRID = "2"
+    AC = "3"
+    HYBRID_3PH = "4"
+    AC_3PH = "6"
+    EMS = "5"
+    GATEWAY = "7"
+    ALL_IN_ONE = "8"
+
+    @classmethod
+    def _missing_(cls, value):
+        """Pick model from the first digit of the device type code."""
+        return cls(value[0])
+
+class Status(IntEnum):
+    """Inverter status."""
+
+    WAITING = 0
+    NORMAL = 1
+    WARNING = 2
+    FAULT = 3
+    FLASHING_FIRMWARE_UPDATE = 4
+    @classmethod
+    def _missing_(cls, value):
+        """Default to 0."""
```

Comment:
There is an unused 

```
class DefaultUnknownIntEnum(IntEnum):
    """Enum that returns unknown instead of blowing up."""
    @classmethod
    def _missing_(cls, value):
        return cls.UNKNOWN  # type: ignore[attr-defined] # must be defined in subclasses because of Enum limits

```
in `__init__.py` which could reduce duplication.  I might be inclined to use -1 for the default unknown / missing value, since that can't appear in an unsigned int16 register. Don't like defaulting to a legal value, since you can't distinguish missing from genuinely 0.

(To be honest I'm not really very big on enums - I tend to just stick to ints for everything !)